pkg/apis/duck/v1/delivery_types.go

```diff
@@ -83,6 +90,17 @@ func (ds *DeliverySpec) Validate(ctx context.Context) *apis.FieldError {
 	return errs
 }
 
+func (ds *DeliverySpec) SetDefaults(ctx context.Context) {
```

Comment:
I'd rather specify these in the upgrade tests configuration rather than by default for everything, until we really get our delivery flakes out.

Reply:
Are you saying that's maybe not what we want to do in the long run? Doing this in the upgrade tests configuration is not going to fix the "product".

Reply:
No, what I think I'm trying to say that until we are confident that we have fixed all the flakes, I would prefer not to add retries that may mask bugs. Hence, adding this so upgrade tests only for now would be my preferred approach.

Reply:
@vaikas I propose the opposite. Set defaults as proposed. But on the other hand create special version of upgrade tests that will have default retries disabled. That additional run will be just informative and will not block PR. Regular upgrade test will run with retries and will block PR if delivery fails.

That should fix the product and have those flakes under control. WDYT?

Reply:
I'm confused what we mean when we say "product". The way I read this PR it's very targeted towards addressing upgrade tests, so by making this default change, all of our e2e / conformance tests will now be doing retries and as we've seen there have been flakes in our tests, so making this change will in my opinion make our tests more likely to mask bugs.
These values also seem very targeted again towards the Upgrade tests.
We have default configmap where should be able to configure these without making this change for all the channels created.
Do these defaults make sense for all the channel types for example?

Reply:
I think this isn't targeting upgrade tests, or any other tests. It targets a valid need that even the in-mem channel should be resilient. It's on most of the examples and docs. If users see it's lacking, that's a first impression ruined.

Reply:
Well, then I guess I'm confused about this commend on L95 :)

Why don't we make it so that these defaults can be configured in our configmaps instead of hardcoded into the defaulting code?

Reply:
+1 for default be written on configmaps, for users to fiddle.

Reply:
+1 also for default on configmaps. The reason I didn't do it is only because it's more work.

Reply:
I chatted with @vaikas and he can do the work to put the default delivery spec in a ConfigMap. Thanks @vaikas !

Reply:
My 2 cents here: I'm not confident to ship our components with retries enabled by default at all, nor in tests nor in releases.

I think shipping a default retry behaviour defines some "opinions" on how you should use the components: user probably doesn't expect retries by default, so he may receive 2 times the same event without figuring out why. Or maybe he could get his/her dispatcher bloated because it's retrying events on and on. Or maybe for the user is fine to lose events, compared to the cost of retries, because he/she's just trying knative in a dev cluster.

Another problem is: which values? I mean, the ones you propose are tailored on the upgrade tests, but my general point here is: how can we find out which values should be default for our users? Does it makes sense to give them some arbitrary default values or is it better if they choose for themselves, testing different values and/or calculating those values based on the use cases?

Although I see the argument of "well behaving/well defined defaults", I still think this is a little bit unexpected. When you install knative today you have a default "predictable" behaviour: no-retry. You want to configure them? Fine, we guide you through with the docs. Configuring retries for a user is really trivial, we can even add more docs for that, just to make sure the user can do it easily.

Reply:
> I think shipping a default retry behaviour defines some "opinions" on how you should use the components: **user probably doesn't expect retries by default**

Maybe this is getting sidetracked from the original thread a bit, but my 2 cents: if a user expects no retries by default then we have a documentation bug. The Eventing specifications are clear that at-least-once delivery is the expected behavior for [Brokers](https://github.com/knative/eventing/blob/master/docs/spec/broker.md#delivery) and [Channels](https://github.com/knative/eventing/blob/master/docs/spec/channel.md#retries).

Reply:
Looking at the specs now:

* For broker spec the definition is definitely unclear, it just says _hey we should have at least once_ yeah but how?
* The channel spec talks about retries, saying that _Channels SHOULD retry resending CloudEvents_ without specifying if this is a default behaviour or an opt-in behaviour, but then it says the _Channels SHOULD support various retry configuration parameters_. So is it opt-in or opt-out? That's not clear to me